### PR TITLE
Add split-debuginfo profile option

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds a new `split-debuginfo` option to Cargo compilation
profiles which gets forwarded to the `-Csplit-debuginfo` codegen option
in rustc. This commit also sets the default, only on macOS, to be
`-Csplit-debuginfo=unpacked`. The purpose of this change is to leverage
rust-lang/rust#79570 to avoid running `dsymutil` on incremental builds
while also preserving a pleasant debugging experience by default. This
should lead to much faster incremental build times on macOS since
`dsymutil` isn't exactly the speediest tool in the world.

This is technically a breaking change in Cargo because we're no longer
by-default producing the `*.dSYM` folders on macOS. If those are still
desired, however, authors can always run `dsymutil` themselves or
otherwise configure `split-debuginfo = 'packed'` in their
manifest/profile configuration.

### CODE REVIEW

**Comment by alexcrichton** [comment]
This will fail CI while I'm still figuring out tests (and waiting on a nightly with support for this option), but I wanted to go ahead and open this earlier to see what others' thoughts on this were. I think that we really do need to change the defaults here despite the possibly breaking change, they can have a massive effect on local development. For comparison Cargo's binary builds in 10.5s with dsymutil enabled (packed debuginfo, historical default), and 2.8s without (and most of that time is spent in the linker). I also think that if we do this right for Linux builds with split-dwarf we'll be able to drive down the size of target directories with less duplicated debuginfo as well.

**Comment by Mark-Simulacrum** [comment]
I am a bit surprised to see this as a profile option rather than something you configure in e.g. .cargo/config - it seems like the kind of thing I would want to set on a system/user level rather than per-project. Am I wrong about that?

I suppose making it a profile option is maximally flexible though, so in that case seems fine, just want to better understand the use case here - e.g., would we expect someone to publish to crates.io with this set?

**Comment by alexcrichton** [comment]
Profile configuration can happen in a number of places nowadays, so by making it a profile option  you can configure it via `.cargo/config.toml` or environment variables in addition to manifest-specific settings.

**Comment by alexcrichton** [comment]
Ok this now has all green tests. The main thing I did was to update the rustc caching to cache failure statuses as well as successful statuses when we learn about rustc args so we can cache the failure to use `-Csplit-debuginfo` on platforms it's not stable or on. Note that even when rustc support rides to stable this is important because the option is only stable on macOS.

**Comment by alexcrichton** [comment]
We discussed this at the Cargo meeting today and the conclusion was that we'll:

* Land the profile configuration option here, don't switch defaults
* Make a post on internals about gathering feedback
* Hopefully switch the default after a release or two

File: src/cargo/ops/cargo_compile.rs [hunk 1]
```diff
@@ -887,19 +882,19 @@ fn generate_targets(
...
             let features = resolved_features.activated_features(pkg.package_id(), features_for);
 
             for kind in requested_kinds {
+                let profile = profiles.get_profile(
+                    pkg.package_id(),
+                    ws.is_member(pkg),
+                    is_local,
+                    unit_for,
+                    target_mode,
+                );
```
Comment (ehuss) [inline src/cargo/ops/cargo_compile.rs:897]: It's not clear to me why this moved inside the loop.

File: src/cargo/core/compiler/standard_lib.rs [hunk 2]
```diff
@@ -162,16 +162,16 @@ pub fn generate_std_roots(
...
         let features = std_features.activated_features(pkg.package_id(), FeaturesFor::NormalOrDev);
 
         for kind in kinds {
+            let profile = profiles.get_profile(
+                pkg.package_id(),
+                /*is_member*/ false,
+                /*is_local*/ false,
+                unit_for,
+                mode,
+            );
```
Comment (ehuss) [inline src/cargo/core/compiler/standard_lib.rs:174]: Same question about why this moved.

File: src/cargo/ops/cargo_compile.rs (hunk 1 above)
Comment (alexcrichton) [reply src/cargo/ops/cargo_compile.rs]: Oh this was a relic of when this was doing a macOS-specific thing internally so the `CompileKind` was passed in, I undid that though so I just undid this as well.

**Comment by ehuss** [comment]
🎉 🎉 🎉 🎉 🎉

@bors r+

**Comment by alexcrichton** [comment]
I've [posted about this on internals](https://internals.rust-lang.org/t/help-test-faster-incremental-debug-macos-builds-on-nightly/14016) for the next step hopefully in turning this on by default.

File: src/doc/src/reference/profiles.md [hunk 3]
```diff
@@ -71,8 +71,26 @@ The valid options are:
...
+#### split-debuginfo
+
+The `split-debuginfo` setting controls the [`-C split-debuginfo` flag] which
+controls whether debug information, if generated, is either placed in the
+executable itself or adjacent to it.
+
+This option is a string and acceptable values are the same as those the
+[compiler accepts][`-C split-debuginfo` flag]. This option is currently not
+passed by default on platforms other than macOS, where it defaults to
+`unpacked`. The default may change in the future for platforms using DWARF
```
Comment (eddyb) [inline src/doc/src/reference/profiles.md:88]: @ehuss looks like this wasn't changed with the rest of the PR?

Comment (eddyb) [reply src/doc/src/reference/profiles.md:88]: Opened #9218 about this.

**Pull request merged**